### ISSUE TITLE
Template network boot configurations (iPXE, GRUB)

### ISSUE THREAD

**Issue by dghubble** [poseidon/matchbox#92]
Allow a grub config or iPXE script file with template elements to be provided directly. Substitute template values from `selector` and/or `metadata`. Currently, a `NetBoot` struct and internal templates determine the iPXE scripts and GRUB configs which are rendered which isn't as customizable.

**Comment by dghubble** [comment]
This is needed for the Kubernetes profiles to be practical via Grub netboot for UEFI. The work-around for now is to write different profiles for each machine which embeds a MAC or UUID in the `coreos.config.url` kernel argument.

**Comment by dghubble** [comment]
Prioritizing iPXE usage right now and improving GRUB use cases next.

**Comment by ropsoft** [comment]
Cool. I was about to open an issue asking if this exists currently! Two questions if I may...:

1.) To make sure I'm not mistaken on the goal here... would this allow for setting metadata to env vars?

```
{
  "id": "coreos-install",
  "name": "CoreOS Install",
  "profile": "install-reboot",
  "metadata": {
    "my_custom_var": "${custom_env_var}"
    "coreos_channel": "alpha",
    "coreos_version": "1109.1.0",
    "ignition_endpoint": "http://bootcfg.foo:8080/ignition",
    "baseurl": "http://bootcfg.foo:8080/assets/coreos"
  }
}
```

2.) Would the env var be evaluated every time the group is matched or just templated in memory to static text a single time when bootcfg starts?

**Comment by dghubble** [comment]
Hm, you mean populating metadata from environment variables from the bootcfg process environment? No, this is more for being able to write an iPXE config or GRUB config yourself and have it be templated with metadata variables, much like the Ignition, Cloud-config, and generic configs today. 

The iPXE and GRUB configs are created from a hardcoded template https://github.com/coreos/coreos-baremetal/blob/master/bootcfg/http/ipxe.go#L17. This has been ok in most cases, but it might be useful to customize that. The "boot" section in a profile is a bit of a special beast, when we might have an advanced mode where a profile can refer to a named iPXE config resource. https://github.com/coreos/coreos-baremetal/blob/master/Documentation/bootcfg.md#profiles

**Comment by SchwarzM** [comment]
it would be nice to see this we could need a hook in this to do additional ipxe scripting before boot.

**Comment by discostur** [comment]
Is there any progress in here? If i try to boot via iPXE i'm getting the following error on the iPXE-Client:

```
Error 0x040ee119
```

In the dnsmasq-docker-image i get the following:

```
dnsmasq-dhcp: 2612948051 available DHCP range: 192.168.67.201 -- 192.168.67.250
dnsmasq-dhcp: 2612948051 vendor class: PXEClient:Arch:00000:UNDI:002001
dnsmasq-dhcp: 2612948051 user class: iPXE
dnsmasq-dhcp: 2612948051 DHCPDISCOVER(eth0) c2:39:19:6d:ba:67 
dnsmasq-dhcp: 2612948051 tags: ipxe, eth0
dnsmasq-dhcp: 2612948051 DHCPOFFER(eth0) 192.168.67.220 c2:39:19:6d:ba:67 
dnsmasq-dhcp: 2612948051 requested options: 1:netmask, 3:router, 6:dns-server, 7:log-server, 
dnsmasq-dhcp: 2612948051 requested options: 12:hostname, 15:domain-name, 17:root-path, 
dnsmasq-dhcp: 2612948051 requested options: 43:vendor-encap, 60:vendor-class, 66:tftp-server, 
dnsmasq-dhcp: 2612948051 requested options: 67:bootfile-name, 119:domain-search, 128, 
dnsmasq-dhcp: 2612948051 requested options: 129, 130, 131, 132, 133, 134, 135, 175, 203
dnsmasq-dhcp: 2612948051 next server: 192.168.67.40
dnsmasq-dhcp: 2612948051 sent size:  1 option: 53 message-type  2
dnsmasq-dhcp: 2612948051 sent size:  4 option: 54 server-identifier  192.168.67.40
dnsmasq-dhcp: 2612948051 sent size:  4 option: 51 lease-time  1h
dnsmasq-dhcp: 2612948051 sent size: 57 option: 67 bootfile-name  http://matchbox01.intern:8080/boot.ipxe
dnsmasq-dhcp: 2612948051 sent size:  4 option: 58 T1  30m
dnsmasq-dhcp: 2612948051 sent size:  4 option: 59 T2  52m30s
dnsmasq-dhcp: 2612948051 sent size:  4 option:  1 netmask  255.255.255.0
dnsmasq-dhcp: 2612948051 sent size:  4 option: 28 broadcast  192.168.67.255
dnsmasq-dhcp: 2612948051 sent size:  4 option:  3 router  192.168.67.40
dnsmasq-dhcp: 2612948051 sent size:  4 option:  6 dns-server  192.168.67.40
```

I think this should be related to 
https://github.com/coreos/matchbox/issues/369 
and this one 
https://github.com/coreos/matchbox/issues/92

**Comment by dghubble** [comment]
We may pick this up soon to allow menus to be used as well. cc @robszumski
